CustomControlLibrary/CurrencyConverterControl.cs

```diff
@@ -0,0 +1,509 @@
+
+        static CurrencyConverterControl()
+        {
+            DefaultStyleKeyProperty.OverrideMetadata(typeof(CurrencyConverterControl), new FrameworkPropertyMetadata(typeof(CurrencyConverterControl)));
+        }
+
+        public CurrencyConverterControl()
+        {
+            InitializeModel();
+            Unloaded += OnUnloaded;
+        }
+
+        private void OnUnloaded(object sender, RoutedEventArgs e)
+        {
+            Unloaded -= OnUnloaded;
+
+            FinalizeUI();
+            FinalizeModel();
+        }
+
+        public override void OnApplyTemplate()
+        {
+            InitializeUI();
+            TryInvalidateExchangeRate();
+            base.OnApplyTemplate();
+        }
+
+        private void InitializeModel()
+        {
+            var resource = Application.GetResourceStream(new Uri("/CustomControlLibrary;component/assets/countries.json", UriKind.Relative));
+            DataContractJsonSerializer ser = new DataContractJsonSerializer(typeof(List<CountryInfo>));
+            var countries = (ser.ReadObject(resource.Stream) as List<CountryInfo>);
+ 
+            SetValue(CounriesItemsKey, countries);
+        }
+
+        private void FinalizeModel()
+        {
+            CounriesItems.Clear();
+            
+            _cancellationTokenSource?.Cancel();
+            _cancellationTokenSource = null;
+        }
+
+        private bool _uiInited;
+        private void InitializeUI()
+        {
+            CurrencyFrom = (ComboBox)GetTemplateChild("PART_CurrencyFrom");
+            CurrencyTo = (ComboBox)GetTemplateChild("PART_CurrencyTo");
+            CurrencyFromValue = (TextBox)GetTemplateChild("PART_CurrencyFromValue") ;
+            CurrencyToValue = (TextBox)GetTemplateChild("PART_CurrencyToValue");
+
+            _uiInited = true;
+        }
+
+        private void FinalizeUI()
+        {
+            CurrencyFrom = null;
+            CurrencyTo = null;
+            CurrencyFromValue = null;
+            CurrencyToValue = null;
+            _uiInited = false;
+        }
+
+        private void InvalidateCurrencyComponentView(ComboBox currencyViewComponent, string currencyId)
+        {
+            var si = currencyViewComponent.SelectedItem as CountryInfo;
+            if (si!=null && si.CurrencyId.Equals(currencyId, StringComparison.OrdinalIgnoreCase))
+            {
+                return;
+            }
+            currencyViewComponent.SelectedIndex = CounriesItems.FindIndex(item => item.CurrencyId.Equals(currencyId, StringComparison.OrdinalIgnoreCase));
+        }
+
+        private async void TryInvalidateExchangeRate()
+        {
+            VisualStateManager.GoToState(this, "SelectorsLoading", true);
+            VisualStateManager.GoToState(this, "InputLoading", true);
+            if (_exchangeRateProvider == null)
+            {
+                // Leave control at loading state while waiting setting of provider.
```

Comment:
you can fire an event here to let user know that you are expecting value of an ExchangeRateProvider.
Something like
OnLoadingExchangeRateProvider(loadingProviderArgs);
if (loadingProviderArgs.Provider != null)
{
   _exchangeRateProvider = loadingProviderArgs.Provider
}